dbms/src/Disks/DiskLocal.cpp

```diff
@@ -0,0 +1,240 @@
+#include "DiskLocal.h"
+#include "DiskFactory.h"
+
+#include <Interpreters/Context.h>
+#include <Common/filesystemHelpers.h>
+#include <Common/quoteString.h>
+
+
+namespace DB
+{
+std::mutex DiskLocal::mutex;
+
+ReservationPtr DiskLocal::reserve(UInt64 bytes) const
+{
+    if (!tryReserve(bytes))
+        return {};
+    return std::make_unique<DiskLocalReservation>(std::static_pointer_cast<const DiskLocal>(shared_from_this()), bytes);
+}
+
+bool DiskLocal::tryReserve(UInt64 bytes) const
+{
+    std::lock_guard lock(mutex);
+    if (bytes == 0)
+    {
+        LOG_DEBUG(&Logger::get("DiskLocal"), "Reserving 0 bytes on disk " << backQuote(name));
+        ++reservation_count;
+        return true;
+    }
+
+    auto available_space = getAvailableSpace();
+    UInt64 unreserved_space = available_space - std::min(available_space, reserved_bytes);
+    if (unreserved_space >= bytes)
+    {
+        LOG_DEBUG(
+            &Logger::get("DiskLocal"),
+            "Reserving " << formatReadableSizeWithBinarySuffix(bytes) << " on disk " << backQuote(name) << ", having unreserved "
+                         << formatReadableSizeWithBinarySuffix(unreserved_space) << ".");
+        ++reservation_count;
+        reserved_bytes += bytes;
+        return true;
+    }
+    return false;
+}
+
+UInt64 DiskLocal::getTotalSpace() const
+{
+    auto fs = getStatVFS(path);
+    UInt64 total_size = fs.f_blocks * fs.f_bsize;
+    if (total_size < keep_free_space_bytes)
+        return 0;
+    return total_size - keep_free_space_bytes;
+}
+
+UInt64 DiskLocal::getAvailableSpace() const
+{
+    /// we use f_bavail, because part of b_free space is
+    /// available for superuser only and for system purposes
+    auto fs = getStatVFS(path);
+    UInt64 total_size = fs.f_bavail * fs.f_bsize;
+    if (total_size < keep_free_space_bytes)
+        return 0;
+    return total_size - keep_free_space_bytes;
+}
+
+UInt64 DiskLocal::getUnreservedSpace() const
+{
+    std::lock_guard lock(mutex);
+    auto available_space = getAvailableSpace();
+    available_space -= std::min(available_space, reserved_bytes);
+    return available_space;
+}
+
+DiskFilePtr DiskLocal::file(const String & path_) const
+{
+    return std::make_shared<DiskLocalFile>(std::static_pointer_cast<const DiskLocal>(shared_from_this()), path_);
```

Comment:
Why we need cast here?

Reply:
Because shared_from_this() return shared_ptr to Space.